File: web/src/api/anisongdb.ts
/**
 * AnisongDB (anisongdb.com) — the database behind Anime Music Quiz and most
 * "AniSong" players. Covers openings, endings AND insert songs; media is
 * served as MP3 / WEBM from naedist.animemusicquiz.com. Key-less, CORS-enabled.
 */
import type { ThemeType, Track } from "@/types";
import { DAY, HOUR, request, rev } from "./http";

const BASE = rev("ipa/moc.bdgnosina//:sptth");
const MEDIA = rev("/moc.ziuqcisumemina.tsidean//:sptth");

interface RawArtist { id: number; names: string[] }
interface RawSong {
  annId: number;
  annSongId: number;
  animeENName: string;
  animeJPName: string;
  animeVintage?: string | null;
  animeType?: string | null;
  songType: string; // "Opening 1" | "Ending 2" | "Insert Song"
  songName: string;
  songArtist: string;
  HQ: string | null;
  MQ: string | null;
  audio: string | null;
  artists?: RawArtist[];
  linked_ids?: Record<string, number | number[] | null | undefined> | null;
}

const FILTERS = { ignore_duplicate: false, opening_filter: true, ending_filter: true, insert_filter: true };

function parseType(songType: string): { type: ThemeType; slug: string; sequence: number | null } {
  const m = /^(Opening|Ending|Insert)\s*(\d+)?/i.exec(songType);
  const kind = m?.[1]?.toLowerCase();
  const n = m?.[2] ? Number(m[2]) : null;
  if (kind === "opening") return { type: "OP", slug: `OP${n ?? ""}`, sequence: n };
  if (kind === "ending") return { type: "ED", slug: `ED${n ?? ""}`, sequence: n };
  return { type: "IN", slug: n ? `IN${n}` : "IN", sequence: n };
}

function first(v: number | number[] | null | undefined): number | null {
  if (Array.isArray(v)) return v[0] ?? null;
  return typeof v === "number" ? v : null;
}

function parseVintage(v: string | null | undefined): { year: number | null; season: Track["anime"]["season"] } {
  if (!v) return { year: null, season: null };
  const m = /(Winter|Spring|Summer|Fall)\s+(\d{4})/i.exec(v);
  if (!m) return { year: null, season: null };
  const s = m[1][0].toUpperCase() + m[1].slice(1).toLowerCase();
  return { year: Number(m[2]), season: s as Track["anime"]["season"] };
}

export function toTrack(s: RawSong): Track | null {
  const audio = s.audio ? MEDIA + s.audio : null;
  const video = s.HQ ? MEDIA + s.HQ : s.MQ ? MEDIA + s.MQ : null;
  if (!audio && !video) return null;
  const { type, slug, sequence } = parseType(s.songType);
  const malId = first(s.linked_ids?.[rev("tsileminaym")]);
  const anilistId = first(s.linked_ids?.[rev("tsilina")]);
  const { year, season } = parseVintage(s.animeVintage);
  return {
    id: `asdb:${s.annSongId}`,
    themeId: -s.annSongId,
    themeSlug: slug,
    type,
    sequence,
    title: s.songName,
    artists: (s.artists?.length ? s.artists : [{ id: 0, names: [s.songArtist] }]).map((a) => ({ id: -(a.id || 0), name: a.names?.[0] ?? s.songArtist, slug: "" })),
    anime: { id: -s.annId, name: s.animeJPName || s.animeENName, slug: malId ? `mal-${malId}` : `ann-${s.annId}`, year, season, malId, anilistId },
    cover: null,
    coverSmall: null,
    audioUrl: audio ?? (video as string),
    videoUrl: video ?? (audio as string),
    resolution: s.HQ ? 720 : s.MQ ? 480 : null,
    tags: s.HQ ? "HQ" : s.MQ ? "MQ" : "",
    version: null,
    episodes: null,
    nsfw: false,
    spoiler: false,
    source: "extra",
  };
}

const dedupe = (list: RawSong[]) => {
  const seen = new Set<number>();
  return list.filter((s) => (seen.has(s.annSongId) ? false : (seen.add(s.annSongId), true)));
};

export async function searchAnisongDB(q: string, signal?: AbortSignal): Promise<Track[]> {
  const body = JSON.stringify({
    anime_search_filter: { search: q, partial_match: true },
    song_name_search_filter: { search: q, partial_match: true },
    artist_search_filter: { search: q, partial_match: true, group_granularity: 0, max_other_artist: 99 },
    and_logic: false,
    ...FILTERS,
  });
  const list = await request<RawSong[]>(`${BASE}/search_request`, { body, fresh: 6 * HOUR, maxAge: 7 * DAY, signal, timeout: 15_000 });
  return dedupe(list ?? []).map(toTrack).filter((t): t is Track => !!t).slice(0, 60);
}

/** All songs for an anime: tries MAL-id lookup, falls back to exact-name search. */
export async function getAnisongsForAnime(malId: number | null | undefined, names: string[], signal?: AbortSignal): Promise<Track[]> {
  let list: RawSong[] = [];
  if (malId) {
    try {
      list = (await request<RawSong[]>(`${BASE}/malIDs_request`, { body: JSON.stringify({ malIds: [malId], ...FILTERS }), fresh: 7 * DAY, maxAge: 30 * DAY, signal, timeout: 15_000 })) ?? [];
    } catch {
      list = [];
    }
  }
  if (!list.length) {
    for (const name of names.filter(Boolean).slice(0, 2)) {
      try {
        const r = await request<RawSong[]>(`${BASE}/search_request`, {
          body: JSON.stringify({ anime_search_filter: { search: name, partial_match: false }, and_logic: false, ...FILTERS }),
          fresh: 7 * DAY,
          maxAge: 30 * DAY,
          signal,
          timeout: 15_000,
        });
        if (r?.length) {
          list = r;
          break;
        }
      } catch {
        /* try next name */
      }
    }
  }
  const tracks = dedupe(list).map(toTrack).filter((t): t is Track => !!t);
  const order = { OP: 0, ED: 1, IN: 2 };
  return tracks.sort((a, b) => order[a.type] - order[b.type] || (a.sequence ?? 0) - (b.sequence ?? 0));
}

export async function getRandomAnisongs(signal?: AbortSignal): Promise<Track[]> {
  const list = await request<RawSong[]>(`${BASE}/get_50_random_songs`, { body: "{}", cacheKey: `${BASE}/get_50_random_songs#${Date.now()}`, noStore: true, signal, timeout: 15_000 });
  return dedupe(list ?? []).map(toTrack).filter((t): t is Track => !!t);
}

File: android/app/src/main/java/com/anibeat/app/data/AnisongDb.java
package com.anibeat.app.data;

import com.anibeat.app.core.Net;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * AnisongDB — вставки и редкие темы (перенос `src/api/anisongdb.ts`).
 * Медиа отдаётся с naedist.animemusicquiz.com (MP3 / WEBM).
 */
public final class AnisongDb {

    private static final String BASE = "https://anisongdb.com/api";
    private static final String MEDIA = "https://naedist.animemusicquiz.com";

    private AnisongDb() {
    }

    private static JSONObject filters() {
        JSONObject o = new JSONObject();
        try {
            o.put("ignore_duplicate", false);
            o.put("opening_filter", true);
            o.put("ending_filter", true);
            o.put("insert_filter", true);
        } catch (Exception ignored) {
        }
        return o;
    }

    private static JSONObject bodyWith(JSONObject extra) {
        JSONObject body = new JSONObject();
        try {
            java.util.Iterator<String> keys = extra.keys();
            while (keys.hasNext()) {
                String k = keys.next();
                body.put(k, extra.get(k));
            }
            java.util.Iterator<String> f = filters().keys();
            while (f.hasNext()) {
                String k = f.next();
                body.put(k, filters().get(k));
            }
        } catch (Exception ignored) {
        }
        return body;
    }

    public static Models.Track toTrack(JSONObject s) {
        try {
            String audioPath = s.isNull("audio") ? null : s.optString("audio", null);
            String hq = s.isNull("HQ") ? null : s.optString("HQ", null);
            String mq = s.isNull("MQ") ? null : s.optString("MQ", null);
            String audio = audioPath != null && !audioPath.isEmpty() ? MEDIA + audioPath : null;
            String video = hq != null && !hq.isEmpty() ? MEDIA + hq : (mq != null && !mq.isEmpty() ? MEDIA + mq : null);
            if (audio == null && video == null) return null;

            String songType = s.optString("songType", "");
            java.util.regex.Matcher m = java.util.regex.Pattern.compile("^(Opening|Ending|Insert)\\s*(\\d+)?", java.util.regex.Pattern.CASE_INSENSITIVE).matcher(songType);
            String type = "IN";
            String slug = "IN";
            Integer sequence = null;
            if (m.find()) {
                String kind = m.group(1).toLowerCase();
                Integer n = m.group(2) != null ? Integer.parseInt(m.group(2)) : null;
                if ("opening".equals(kind)) {
                    type = "OP";
                    slug = "OP" + (n == null ? "" : n);
                    sequence = n;
                } else if ("ending".equals(kind)) {
                    type = "ED";
                    slug = "ED" + (n == null ? "" : n);
                    sequence = n;
                } else {
                    type = "IN";
                    slug = n == null ? "IN" : "IN" + n;
                    sequence = n;
                }
            }

            Models.Track t = new Models.Track();
            int annSongId = s.optInt("annSongId");
            t.id = "asdb:" + annSongId;
            t.themeId = -annSongId;
            t.themeSlug = slug;
            t.type = type;
            t.sequence = sequence;
            t.title = s.optString("songName", slug);
            JSONArray artists = s.optJSONArray("artists");
            if (artists != null && artists.length() > 0) {
                for (int i = 0; i < artists.length(); i++) {
                    JSONObject a = artists.optJSONObject(i);
                    if (a == null) continue;
                    Models.ArtistRef ref = new Models.ArtistRef();
                    ref.id = -a.optInt("id");
                    JSONArray names = a.optJSONArray("names");
                    ref.name = names != null && names.length() > 0 ? names.optString(0) : s.optString("songArtist", "");
                    ref.slug = "";
                    t.artists.add(ref);
                }
            } else {
                Models.ArtistRef ref = new Models.ArtistRef();
                ref.id = 0;
                ref.name = s.optString("songArtist", "Неизвестный исполнитель");
                ref.slug = "";
                t.artists.add(ref);
            }

            Models.AnimeRef anime = new Models.AnimeRef();
            int annId = s.optInt("annId");
            anime.id = -annId;
            String jp = s.optString("animeJPName", "");
            String en = s.optString("animeENName", "");
            anime.name = jp != null && !jp.isEmpty() ? jp : en;
            Object linked = s.opt("linked_ids");
            Integer malId = null;
            Integer anilistId = null;
            if (linked instanceof JSONObject) {
                JSONObject l = (JSONObject) linked;
                malId = first(l.opt("myanimelist"));
                anilistId = first(l.opt("anilist"));
            }
            anime.malId = malId;
            anime.anilistId = anilistId;
            anime.slug = malId != null ? "mal-" + malId : "ann-" + annId;
            String vintage = s.isNull("animeVintage") ? null : s.optString("animeVintage", null);
            if (vintage != null) {
                java.util.regex.Matcher vm = java.util.regex.Pattern.compile("(Winter|Spring|Summer|Fall)\\s+(\\d{4})", java.util.regex.Pattern.CASE_INSENSITIVE).matcher(vintage);
                if (vm.find()) {
                    String season = vm.group(1);
                    season = season.substring(0, 1).toUpperCase() + season.substring(1).toLowerCase();
                    anime.season = season;
                    anime.year = Integer.parseInt(vm.group(2));
                }
            }
            t.anime = anime;

            // Музыка — только аудио: видео-ссылку в аудио не подставляем.
            t.audioUrl = audio != null ? audio : "";
            t.videoUrl = video != null ? video : "";
            t.resolution = hq != null && !hq.isEmpty() ? 720 : (mq != null && !mq.isEmpty() ? 480 : null);
            t.tags = hq != null && !hq.isEmpty() ? "HQ" : (mq != null && !mq.isEmpty() ? "MQ" : "");
            t.source = "extra";
            return t;
        } catch (Exception e) {
            return null;
        }
    }

    private static Integer first(Object value) {
        if (value instanceof JSONArray) {
            JSONArray arr = (JSONArray) value;
            return arr.length() > 0 ? arr.optInt(0) : null;
        }
        if (value instanceof Number) return ((Number) value).intValue();
        return null;
    }

    private static List<Models.Track> mapAll(JSONArray array) {
        List<Models.Track> out = new ArrayList<>();
        Set<Integer> seen = new HashSet<>();
        if (array == null) return out;
        for (int i = 0; i < array.length(); i++) {
            JSONObject o = array.optJSONObject(i);
            if (o == null) continue;
            int id = o.optInt("annSongId");
            if (!seen.add(id)) continue;
            Models.Track t = toTrack(o);
            if (t != null) out.add(t);
            if (out.size() >= 60) break;
        }
        return out;
    }

    /** Поиск по базе (расширенный режим поиска). */
    public static void search(String query, Api.Cb<List<Models.Track>> cb) {
        JSONObject params = new JSONObject();
        try {
            JSONObject anime = new JSONObject();
            anime.put("search", query);
            anime.put("partial_match", true);
            JSONObject song = new JSONObject();
            song.put("search", query);
            song.put("partial_match", true);
            JSONObject artist = new JSONObject();
            artist.put("search", query);
            artist.put("partial_match", true);
            artist.put("group_granularity", 0);
            artist.put("max_other_artist", 99);
            params.put("anime_search_filter", anime);
            params.put("song_name_search_filter", song);
            params.put("artist_search_filter", artist);
            params.put("and_logic", false);
        } catch (Exception ignored) {
        }
        Net.post(BASE + "/search_request", bodyWith(params).toString(), (json, error) -> {
            if (json == null) {
                cb.on(new ArrayList<>(), error);
                return;
            }
            JSONArray array = json.optJSONArray("array");
            cb.on(mapAll(array), error);
        });
    }

    /** Все песни аниме: сначала по MAL id, затем точным совпадением имени. */
    public static void forAnime(Integer malId, List<String> names, Api.Cb<List<Models.Track>> cb) {
        if (malId != null) {
            JSONObject body = new JSONObject();
            try {
                JSONArray ids = new JSONArray();
                ids.put(malId);
                body.put("malIds", ids);
            } catch (Exception ignored) {
            }
            Net.post(BASE + "/malIDs_request", bodyWith(body).toString(), (json, error) -> {
                JSONArray array = json == null ? null : json.optJSONArray("array");
                List<Models.Track> tracks = mapAll(array);
                if (!tracks.isEmpty()) {
                    sortInPlace(tracks);
                    cb.on(tracks, error);
                } else {
                    byNames(names, 0, cb);
                }
            });
        } else {
            byNames(names, 0, cb);
        }
    }

    private static void byNames(List<String> names, int index, Api.Cb<List<Models.Track>> cb) {
        if (names == null || index >= Math.min(2, names.size())) {
            cb.on(new ArrayList<>(), null);
            return;
        }
        String name = names.get(index);
        if (name == null || name.isEmpty()) {
            byNames(names, index + 1, cb);
            return;
        }
        JSONObject params = new JSONObject();
        try {
            JSONObject anime = new JSONObject();
            anime.put("search", name);
            anime.put("partial_match", false);
            params.put("anime_search_filter", anime);
            params.put("and_logic", false);
        } catch (Exception ignored) {
        }
        Net.post(BASE + "/search_request", bodyWith(params).toString(), (json, error) -> {
            JSONArray array = json == null ? null : json.optJSONArray("array");
            List<Models.Track> tracks = mapAll(array);
            if (!tracks.isEmpty()) {
                sortInPlace(tracks);
                cb.on(tracks, error);
            } else {
                byNames(names, index + 1, cb);
            }
        });
    }

    private static void sortInPlace(List<Models.Track> tracks) {
        tracks.sort((a, b) -> {
            int oa = order(a.type);
            int ob = order(b.type);
            if (oa != ob) return oa - ob;
            int sa = a.sequence == null ? 0 : a.sequence;
            int sb = b.sequence == null ? 0 : b.sequence;
            return sa - sb;
        });
    }

    private static int order(String type) {
        if ("OP".equals(type)) return 0;
        if ("ED".equals(type)) return 1;
        return 2;
    }

    /** 50 случайных песен — для «Радио». */
    public static void random(Api.Cb<List<Models.Track>> cb) {
        Net.post(BASE + "/get_50_random_songs", "{}", (json, error) -> {
            JSONArray array = json == null ? null : json.optJSONArray("array");
            cb.on(mapAll(array), error);
        });
    }
}
